typing/typecore.ml

```diff
@@ -2147,24 +2125,19 @@ and type_pat_aux
       generalize_structure ty;
       let ty, expected_ty' = instance ty, ty in
       unify_pat_types ~refine loc env ty (instance expected_ty);
+      let expected_ty' =
+        match repr expected_ty' with
+        | {desc = Tpoly (ty, tl)} -> snd (instance_poly ~keep_names:true false tl ty)
+        | _ -> expected_ty'
+      in
       type_pat category sp' expected_ty' (fun p ->
         (*Format.printf "%a@.%a@."
           Printtyp.raw_type_expr ty
           Printtyp.raw_type_expr p.pat_type;*)
         pattern_force := force :: !pattern_force;
         let extra = (Tpat_constraint cty, loc, sp'.ppat_attributes) in
         let p : k general_pattern =
-          match category, (p : k general_pattern) with
-          | Value, {pat_desc = Tpat_var (id,s); _} ->
-            {p with
-              pat_type = ty;
-              pat_desc =
-                Tpat_alias
-                  ({p with pat_desc = Tpat_any; pat_attributes = []}, id,s);
-              pat_extra = [extra];
-            }
-          | _, p ->
-             { p with pat_type = ty; pat_extra = extra::p.pat_extra }
```

Comment:
There is some weird encoding going on here in the old code that I don't understand. Could you explain the change?

Reply:
I can't quite explain what the old code was doing, but it relates to the fact that `type_pat` did not used to track levels and sharing properly. Now it does so this can be deleted.